### PR TITLE
PHPC-2111: Replace zend_parse_parameters calls with macros

### PR DESCRIPTION
**Pull request by alcaeus**
PHPC-2111

### CODE REVIEW

File: src/MongoDB/Manager.c [hunk 1]
```diff
@@ -257,12 +256,12 @@ static PHP_METHOD(Manager, __construct)
...
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|s!a/!a/!", &uri_string, &uri_string_len, &options, &driverOptions) == FAILURE) {
-		zend_restore_error_handling(&error_handling);
-		return;
-	}
-	zend_restore_error_handling(&error_handling);
+	PHONGO_PARSE_PARAMETERS_START(0, 3)
+	Z_PARAM_OPTIONAL
+	Z_PARAM_STRING_OR_NULL(uri_string, uri_string_len)
+	Z_PARAM_ARRAY_EX(options, 1, 1)
+	Z_PARAM_ARRAY_EX(driverOptions, 1, 1)
```
Comment (jmikola) [inline src/MongoDB/Manager.c:263]: Noted that `Z_PARAM_ARRAY_EX`'s second and third arguments correspond to `check_null` and `separate`, respectively (see: [lxr](https://heap.space/xref/PHP-8.1/Zend/zend_API.h?r=76fcd70c#1496)).

File: src/MongoDB/ReadPreference.c [hunk 2]
```diff
@@ -194,12 +193,12 @@ static PHP_METHOD(ReadPreference, __construct)
...
-	zend_replace_error_handling(EH_THROW, phongo_exception_from_phongo_domain(PHONGO_ERROR_INVALID_ARGUMENT), &error_handling);
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "z|a/!a!", &mode, &tagSets, &options) == FAILURE) {
-		zend_restore_error_handling(&error_handling);
-		return;
-	}
-	zend_restore_error_handling(&error_handling);
+	PHONGO_PARSE_PARAMETERS_START(1, 3)
+	Z_PARAM_ZVAL(mode)
+	Z_PARAM_OPTIONAL
+	Z_PARAM_ARRAY_EX(tagSets, 1, 1)
```
Comment (jmikola) [inline src/MongoDB/ReadPreference.c:199]: Noted this applies `separate` for the parsed zval. I believe the Manager and ReadPreference constructors were the only place where this was necessary.

File: src/MongoDB/WriteConcern.c [hunk 3]
```diff
@@ -145,7 +144,7 @@ static PHP_METHOD(WriteConcern, __construct)
 
 	switch (ZEND_NUM_ARGS()) {
 		case 3:
-			if (Z_TYPE_P(journal) != IS_NULL) {
+			if (journal && Z_TYPE_P(journal) != IS_NULL) {
```
Comment (jmikola) [inline src/MongoDB/WriteConcern.c:147]: Why was this change necessary? Did `zend_parse_parameters` previously _always_ initialize `zval *journal` or what this a long-standing bug?

**Review by jmikola — APPROVED** [review]

File: src/BSON/Binary.c [hunk 4]
```diff
@@ -94,20 +94,17 @@ static HashTable* php_phongo_binary_get_properties_hash(phongo_compat_object_han
...
-	zend_replace_error_handling(EH_THROW, phongo_exception_from_phongo_domain(PHONGO_ERROR_INVALID_ARGUMENT), &error_handling);
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "sl", &data, &data_len, &type) == FAILURE) {
-		zend_restore_error_handling(&error_handling);
-		return;
-	}
-	zend_restore_error_handling(&error_handling);
+	PHONGO_PARSE_PARAMETERS_START(2, 2)
+	Z_PARAM_STRING(data, data_len)
+	Z_PARAM_LONG(type)
+	PHONGO_PARSE_PARAMETERS_END();
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/BSON/Binary.c:107]: Yes, my apologies for creating a duplicate. I was searching jira for tickets but it didn't show me that ticket; looks like I need to up my ticket-finding game. I closed PHPC-1656 as a duplicate.

File: src/MongoDB/WriteConcern.c (hunk 3 above)
Comment (alcaeus) [reply src/MongoDB/WriteConcern.c:147]: Without the additional check I was presented with a compiler warning about `journal` potentially being uninitialised. Looking through the sources, we can see that [`zend_parse_arg_zval_deref`](https://heap.space/xref/PHP-8.1/Zend/zend_API.h?r=76fcd70c#2256) will always assign the `arg` pointer to `*dest` if `check_null` is false (as is the case for `journal`). I'm not entirely sure what `arg` would be for optional arguments that weren't provided, or why this wasn't an issue before, especially considering that the call stack always ends up in the above `zend_parse_arg_zval_deref` function. I lack the internals knowledge to figure that one out, hence the simple `NULL` check.
